### PR TITLE
Prioritize history vs optimizer result

### LINKED ISSUE #327: Mismatch of history vs optimizer reported optimal values
When executing the rosenbrock.ipynb notebook on develop branch, I get massive complaints:

```
Function values from history and optimizer do not match: 1.3168387678656086, 2.0179911928533514
Parameters obtained from history and optimizer do not match: [0.99553374 0.98995083 0.97544456 0.94768476 0.89550744 0.8019584
 0.64185456 0.40180156 0.14794723 0.00844711], [0.9940831  0.98497745 0.96635783 0.93780681 0.87380595 0.74153588
 0.53395112 0.25523609 0.03957006 0.00100739]
Function values from history and optimizer do not match: 1.6349167203078765, 2.336760892779954
```

I guess something must still be wrong about history values vs. values returned by optimizers. Happens for various optimizers.

### LINKED ISSUE #743: Flaky test_result_from_hdf5_history?
Fails here https://github.com/ICB-DCM/pyPESTO/runs/3893112578?check_suite_focus=true

I don't think that's related to that commit.

```
________________________ test_result_from_hdf5_history _________________________

    def test_result_from_hdf5_history():
        problem = create_petab_problem()
        with tempfile.TemporaryDirectory(dir=".") as tmpdirname:
            _, fn = tempfile.mkstemp(".hdf5", dir=f"{tmpdirname}")
    
            history_options_hdf5 = pypesto.HistoryOptions(trace_record=True,
                                                          storage_file=fn)
            # optimize with history saved to hdf5
            result = pypesto.optimize.minimize(
                problem=problem, n_starts=1,
                history_options=history_options_hdf5)
    
            result_from_hdf5 = optimization_result_from_history(fn)
    
            # Currently 'exitflag', 'time' and 'message' are not loaded.
            arguments = [ID, X, FVAL, GRAD, HESS, RES, SRES,
                         N_FVAL, N_GRAD, N_HESS, N_RES, N_SRES, X0, FVAL0]
            for key in arguments:
                if result.optimize_result.list[0][key] is None:
                    assert result_from_hdf5.optimize_result.list[0][key] is None
                elif isinstance(result.optimize_result.list[0][key], np.ndarray):
>                   np.testing.assert_almost_equal(
                        result.optimize_result.list[0][key],
                        result_from_hdf5.optimize_result.list[0][key]
                    )
E                   AssertionError: 
E                   Arrays are not almost equal to 7 decimals
E                   
E                   Mismatched elements: 2 / 2 (100%)
E                   Max absolute difference: 0.0001968
E                   Max relative difference: 0.00014251
E                    x: array([-0.915745 , -7.9470828])
E                    y: array([-0.9156145, -7.946886 ])

test/base/test_store.py:424: AssertionError
```

### PR DESCRIPTION
**Pull request by yannikschaelte**
* check admissibility in optimizer history min update
* implement history-result priorization
* in `OptimizerHistory._compute_vals_from_trace`, scan all entries for derivatives, instead of only +1.
* Fixed issue of profiles when gradients are not returned from optimization.

fixes #327
hopefully also fixes #743 (lots of things refactored)

### CODE REVIEW

**Comment by yannikschaelte** [comment]
@FFroehlich @jvanhoefer @dweindl would [this](https://github.com/ICB-DCM/pyPESTO/pull/775/files#diff-72259b91efe93bb58c38a09cf0d86a1eece92e7c3cc5426ee125c6ba2d8ab6cfR198) be the intended logic for #327?

Or e.g. also overwrite gradient etc from history if `np.isclose(result.fval, history.fval_min` if e.g. `history.grad_min is not None`?

**Comment by yannikschaelte** [comment]
Alternative: Only ever record values in the history trace that fulfill the bounds.

**Comment by dweindl** [comment]
> Alternative: Only ever record values in the history trace that fulfill the bounds.

Optionally, but as default?

**Comment by dweindl** [comment]
> would [...] be the intended logic for #327? 

yes

> Or e.g. also overwrite gradient etc from history if `np.isclose(result.fval, history.fval_min` if e.g. `history.grad_min is not None`?

Not sure I got that. We want to ensure that`{fval,grad,x}_min` always match. If we replace one, then also the others.

File: pypesto/optimize/optimizer.py [hunk 1]
```diff
@@ -159,55 +198,56 @@ def wrapped_minimize(self, problem, x0, id, allow_failed_starts,
...
+
+    Returns
+    -------
+    result: The in-place modified result.
+    """
+    if optimize_options is None:
+        optimize_options = OptimizeOptions()
+
+    # logging
+    history_fval, result_fval = optimizer_history.fval_min, result.fval
```
Comment (FFroehlich) [inline pypesto/optimize/optimizer.py:222]: no more warning if parameters do not match?

**Review by FFroehlich — APPROVED** [review]
modify `get_trimmed_indices` to only return indices for admissable points?

**Comment by FFroehlich** [comment]
> @FFroehlich @jvanhoefer @dweindl would [this](https://github.com/ICB-DCM/pyPESTO/pull/775/files#diff-72259b91efe93bb58c38a09cf0d86a1eece92e7c3cc5426ee125c6ba2d8ab6cfR198) be the intended logic for #327? Then happy to add tests+fix.

looks good!

> Or e.g. also overwrite gradient etc from history if `np.isclose(result.fval, history.fval_min` if e.g. `history.grad_min is not None`?

I think that makes sense, especially for least-squares solvers. But I agree with Daniel here, if `np.isclose(result.fval, history.fval_min)`, replace all relevant keys with values from history if anything is not available, not just the unavailable ones.

**Comment by yannikschaelte** [comment]
> > Or e.g. also overwrite gradient etc from history if `np.isclose(result.fval, history.fval_min` if e.g. `history.grad_min is not None`?
> 
> Not sure I got that. We want to ensure that`{fval,grad,x}_min` always match. If we replace one, then also the others.

By that I meant that if `history_beats_optimizer is False`, then, given np.isclose(...), do we still want to use history values to be added into the result, if the history has values recorded that the optimizer did not return. From @FFroehlich I infer yes.

File: pypesto/optimize/optimizer.py (hunk 1 above)
Comment (yannikschaelte) [reply pypesto/optimize/optimizer.py:222]: will add (as logging too, right?)

**Comment by yannikschaelte** [comment]
@FFroehlich @dweindl fixed a variety of things, now review welcome :)

File: test/base/test_history.py [hunk 2]
```diff
@@ -80,45 +85,45 @@ def check_history(self):
...
                             self.problem, self.history_options,
                             n_starts=n_starts,
                         )
 
-    def check_load_from_file(self, start: OptimizerResult, id: str):
-        """Verify we can reconstitute OptimizerResult from csv file"""
-
+    def check_load_from_file(self, start: optimize.OptimizerResult, id: str):
+        """Verify we can reconstitute OptimizerResult from history file"""
+        # TODO other implementations
```
Comment (FFroehlich) [inline test/base/test_history.py:105]: which are missing here?

**Review by FFroehlich — APPROVED** [review]

File: test/base/test_history.py (hunk 2 above)
Comment (yannikschaelte) [reply test/base/test_history.py:105]: I guess just memory now

File: pypesto/objective/history.py [hunk 3]
```diff
@@ -1329,44 +1343,75 @@ def _compute_vals_from_trace(self):
...
+        for var in ['res', 'grad', 'sres', 'hess']:
+            if not getattr(self.history.options, f'trace_record_{var}'):
+                continue  # var not saved in history
+            # first try index of optimal function value
+            if self.extract_from_history(var, ix_min):
+                continue
+            # gradients may be evaluated at different indices, therefore
+            #  iterate over all and check whether any has the same parameter
+            #  and the desired field filled
+            # for res we do the same because otherwise randomly None
```
[the remark this answers is not in the document]
Comment (yannikschaelte) [reply pypesto/objective/history.py:1374]: could not figure out why. test_history.py::ResModeHistoryTest.test_trace_all failed in 20% of cases

File: test/base/test_history.py (hunk 2 above)
Comment (FFroehlich) [reply test/base/test_history.py:105]: but memory, by definition, is not saved to a file. I think we will have a hard time ever implementing this ;)

**Pull request merged**